src/Common/tests/System/Net/VirtualNetwork/VirtualNetwork.cs

```diff
@@ -36,7 +36,24 @@ public void ReadFrame(bool server, out byte[] buffer)
 
             semaphore.Wait(WaitForReadDataTimeoutMilliseconds);
 
-            bool dequeueSucceeded = packetQueue.TryDequeue(out buffer);
+            bool dequeueSucceeded = false;
+            int remainingRetries = 3;
+            int backOffDelayMilliseconds = 10;
+
+            do
+            {
+                dequeueSucceeded = packetQueue.TryDequeue(out buffer);
```

Comment:
I now see what you mean. The reason I didn't think that was possible because the tests were executing in <10s (about 7-8sec) and the semaphore wait timeout was 10s. 

Retrying on the semaphore wait doesn't make the code better as the idea was to eventually time out.
Increasing the timeout to 30sec and throwing a TimeoutException if the semaphore wait returns false.